### PR TITLE
Template help group param merging

### PR DESCRIPTION
**Pull request by seancpeters**
Better matching across templates with the same group identity.
Appended ".2.0" to the identity of all the 2.0 templates.

### CODE REVIEW

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 1]
```diff
@@ -118,33 +118,41 @@ public IFilteredTemplateInfo UnambiguousTemplateToUse
...
-                            if (match == null)
-                            {
-                                match = info;
-                            }
-                            else
-                            {
-                                return UseSecondaryCriteriaToDisambiguateTemplateMatches();
-                            }
-                        }
+                        return _unambiguousTemplateToUse = languageMatchedTemplates.First();
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: languageMatchedTemplates is an IReadOnlyList<T>, so we can just use `[0]` to get the first item

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 2]
```diff
@@ -118,33 +118,41 @@ public IFilteredTemplateInfo UnambiguousTemplateToUse
...
 
                 return UseSecondaryCriteriaToDisambiguateTemplateMatches();
             }
         }
 
+        private static IReadOnlyList<IFilteredTemplateInfo> FindLanguageMatchedTemplates(IEnumerable<IFilteredTemplateInfo> listToFilter, string language)
+        {
+            List<IFilteredTemplateInfo> languageMatches = new List<IFilteredTemplateInfo>();
+
+            if (string.IsNullOrEmpty(language))
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Wouldn't this be the other way around? If no language is specified, everything matches?

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 3]
```diff
@@ -168,6 +176,14 @@ private IFilteredTemplateInfo UseSecondaryCriteriaToDisambiguateTemplateMatches(
                     {
                         _secondaryFilteredUnambiguousTemplateToUse = matchesIncludingParameterChecks.First();
                     }
+                    else if (EnvironmentSettings.Host.TryGetHostParamDefault("prefs:language", out string defaultLanguage))
+                    {
+                        IReadOnlyList<IFilteredTemplateInfo> languageFiltered = FindLanguageMatchedTemplates(matchesIncludingParameterChecks, defaultLanguage);
+                        if (languageFiltered.Count == 1)
+                        {
+                            _secondaryFilteredUnambiguousTemplateToUse = languageFiltered.First();
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: .First() again

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs (hunk 2 above)
Comment (seancpeters) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: The method is used in contexts that want to find explicit language matches.

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs (hunk 1 above)
Comment (seancpeters) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: 👍

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 4]
```diff
@@ -113,65 +146,90 @@ public IFilteredTemplateInfo UnambiguousTemplateToUse
...
+            if (_secondaryFilteredUnambiguousTemplateGroupToUse == null)
             {
                 if (_matchedTemplatesWithSecondaryMatchInfo == null)
                 {
                     _matchedTemplatesWithSecondaryMatchInfo = FilterTemplatesOnParameters(_matchedTemplates).Where(x => x.IsMatch).ToList();
 
-                    IReadOnlyList<IFilteredTemplateInfo> matchesIncludingParameterChecks = _matchedTemplatesWithSecondaryMatchInfo.Where(x => x.IsParameterMatch).ToList();
+                    IReadOnlyList<IFilteredTemplateInfo> matchesAfterParameterChecks = _matchedTemplatesWithSecondaryMatchInfo.Where(x => x.IsParameterMatch).ToList();
+
+                    if (!matchesAfterParameterChecks.Any())
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: We can just check the count of the list here

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 5]
```diff
@@ -362,7 +428,11 @@ private IReadOnlyList<ITemplateInfo> TemplatesToDisplayInfoAbout
             {
                 IEnumerable<ITemplateInfo> templateList;
 
-                if (_matchedTemplatesWithSecondaryMatchInfo != null && _matchedTemplatesWithSecondaryMatchInfo.Count > 0)
+                if (UnambiguousTemplateGroupToUse != null && UnambiguousTemplateGroupToUse.Any())
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Can check count

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 6]
```diff
@@ -615,14 +780,15 @@ private CreationResultStatus EnterMaintenanceFlow()
         {
             PerformCoreTemplateQuery();
 
-            if (UnambiguousTemplateToUse != null)
+            if (UnambiguousTemplateGroupToUse != null && UnambiguousTemplateGroupToUse.Any())
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Count

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 7]
```diff
@@ -97,13 +96,47 @@ public New3Command(string commandName, ITemplateEngineHost host, Action<IEngineE
 
         public bool SkipUpdateCheck => _app.InternalParamHasValue("--skip-update-check");
 
-        public IFilteredTemplateInfo UnambiguousTemplateToUse
+        private static bool AreAllTemplatesSameGroupIdentity(IEnumerable<IFilteredTemplateInfo> templateList)
+        {
+            Func<IFilteredTemplateInfo, string> groupIdentitySelector = (x) => x.Info.GroupIdentity;
+            return templateList.AllAreTheSame(groupIdentitySelector, StringComparer.OrdinalIgnoreCase);
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Why break out the func to its own line?

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 8]
```diff
@@ -97,13 +96,47 @@ public New3Command(string commandName, ITemplateEngineHost host, Action<IEngineE
...
+        }
+
+        private static IFilteredTemplateInfo HighestPrecedenceTemplateIfAllSameGroupIdentity(IReadOnlyList<IFilteredTemplateInfo> templateList)
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Action names for methods

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 9]
```diff
@@ -323,6 +382,13 @@ private string GetTemplateParameterErrorsMessage(ITemplateInfo templateInfo, out
             return null;
         }
 
+        private string GetLanguageMismatchErrorMessage(string inputLanguage)
+        {
+            string invalidLanguageErrorText = LocalizableStrings.InvalidTemplateParameterValues;
+            invalidLanguageErrorText += Environment.NewLine + string.Format(LocalizableStrings.InvalidParameterDetail, "-lang", inputLanguage, "--language");
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Let's use full switch names when mentioning them in output messages (unless we're using the user specified form)

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 10]
```diff
@@ -390,27 +460,34 @@ private void DisplayTemplateList(bool showAll = false)
 
             foreach (IGrouping<string, ITemplateInfo> grouping in grouped)
             {
-                List<string> languages = new List<string>();
+                List<string> languageForDisplay = new List<string>();
+                HashSet<string> uniqueLanguages = new HashSet<string>();
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Should set a comparer

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 11]
```diff
@@ -390,27 +460,34 @@ private void DisplayTemplateList(bool showAll = false)
...
 
                 foreach (ITemplateInfo info in grouping)
                 {
                     if (info.Tags != null && info.Tags.TryGetValue("language", out ICacheTag languageTag))
                     {
                         foreach (string lang in languageTag.ChoicesAndDescriptions.Keys)
                         {
-                            if (string.IsNullOrEmpty(Language) && string.Equals(defaultLanguage, lang, StringComparison.OrdinalIgnoreCase))
+                            if (!uniqueLanguages.Contains(lang))
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Couldn't this be converted to `if(uniqueLanuages.Add(lang)` and the call to add at the end of the block be removed?

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 12]
```diff
@@ -390,27 +460,34 @@ private void DisplayTemplateList(bool showAll = false)
...
+                                {
+                                    languageForDisplay.Add($"[{lang}]");
+                                }
+                                else
+                                {
+                                    languageForDisplay.Add(lang);
+                                    uniqueLanguages.Add(lang);
+                                }
+
+                                uniqueLanguages.Add(lang);
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: This add happens both in the else and outside of it - the one in the else can probably be removed, as could this one if the initial contains check is replaced with an Add

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs [hunk 13]
```diff
@@ -521,38 +612,123 @@ private CreationResultStatus EnterMaintenanceFlow()
...
                 }
+            }
 
-                DisplayTemplateList();
-                return CreationResultStatus.Success;
+            DisplayTemplateList();
+            return CreationResultStatus.Success;
+        }
+
+        private CreationResultStatus SingularGroupDisplayTemplateHelp()
```
Comment (mlorbetske) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Not sure how to read this method name, not sure if it reflects an action or not

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs (hunk 4 above)
Comment (seancpeters) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: 👍

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs (hunk 7 above)
Comment (seancpeters) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Seems more readable to me, but I don't feel particularly strongly about it either way.

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs (hunk 8 above)
Comment (seancpeters) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Changing to FindHighestPrecedenceTemplateIfAllSameGroupIdentity

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs (hunk 9 above)
Comment (seancpeters) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: I did it this way to be more consistent with the output from GetTemplateParameterErrorsMessage(), which lists the canonical (full name) at the end of the message. Perhaps a quick discussion offline is best here.

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs (hunk 10 above)
Comment (seancpeters) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: 👍

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs (hunk 12 above)
Comment (seancpeters) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Yeah, this is a mess, fixing.

File: src/Microsoft.TemplateEngine.Cli/New3Command.cs (hunk 13 above)
Comment (seancpeters) [inline src/Microsoft.TemplateEngine.Cli/New3Command.cs]: Changing to DisplayTemplateHelpForSingularGroup

**Pull request merged**